pkg/parser/external/java_parser.go

```diff
@@ -34,56 +32,69 @@ func newJavaParser(typ string) *javaParser {
 	return &javaParser{typ: typ}
 }
 
-type parseResult struct {
-	Statements []string `json:"statements"`
-	Position   int      `json:"position"`
-	Error      string   `json:"error"`
+func getServerAddress() string {
+	ip := getEnv("SQLFLOW_EXTERNAL_PARSER_IP", "localhost")
+	port := getEnv("SQLFLOW_EXTERNAL_PARSER_PORT", "12345")
+	return fmt.Sprintf("%s:%s", ip, port)
 }
 
-func (p *javaParser) Dialect() string {
-	return "java"
-}
-
-func (p *javaParser) Parse(program string) ([]string, int, error) {
-	// cwd is used to store train scripts and save output models.
-	cwd, err := ioutil.TempDir("/tmp", "sqlflow")
-	if err != nil {
-		return nil, -1, err
+func isServerUp(address string) bool {
+	cmd := exec.Command("curl", "-v", address)
+	if err := cmd.Run(); err != nil {
+		return false
 	}
-	defer os.RemoveAll(cwd)
+	return true
+}
 
-	inputFile := filepath.Join(cwd, "input.sql")
-	outputFile := filepath.Join(cwd, "output.json")
-	if err := ioutil.WriteFile(inputFile, []byte(program), 0755); err != nil {
-		return nil, -1, err
+func startServerIfNotUp() error {
+	address := getServerAddress()
+	if isServerUp(address) {
+		return nil
 	}
 
-	// TODO(yi): It is very expensive to start a Java process.  It
-	// slows down SQLFlow server's QPS if for every parsing
-	// operation, we'd have to start a Java process.
+	port := getEnv("SQLFLOW_EXTERNAL_PARSER_PORT", "12345")
 	cmd := exec.Command("java",
 		"-cp", "/opt/sqlflow/parser/parser-1.0-SNAPSHOT-jar-with-dependencies.jar",
-		"org.sqlflow.parser.ParserAdaptorCmd",
-		"-p", p.typ,
-		"-i", inputFile,
-		"-o", outputFile)
-	if output, err := cmd.CombinedOutput(); err != nil {
-		return nil, -1, fmt.Errorf("%s %v", output, err)
+		"org.sqlflow.parser.ParserGrpcServer",
+		"-p", port)
+	if err := cmd.Start(); err != nil {
+		return err
 	}
 
-	output, err := ioutil.ReadFile(outputFile)
-	if err != nil {
+	for i := 0; i < 3; i++ {
+		time.Sleep(time.Second)
+		if isServerUp(address) {
+			return nil
+		}
+	}
+
+	return fmt.Errorf("unable to start external parser service")
+}
+
+func (p *javaParser) Parse(program string) ([]string, int, error) {
+	if err := startServerIfNotUp(); err != nil {
 		return nil, -1, err
 	}
 
-	var pr parseResult
-	if err = json.Unmarshal(output, &pr); err != nil {
+	c, err := grpc.Dial(getServerAddress(), grpc.WithInsecure())
+	if err != nil {
 		return nil, -1, err
 	}
+	defer c.Close()
 
-	if pr.Error != "" {
-		return nil, -1, fmt.Errorf(pr.Error)
+	r, err := proto.NewParserClient(c).Parse(context.Background(), &proto.ParserRequest{Dialect: p.typ, SqlProgram: program})
+	if err != nil {
+		return nil, -1, err
+	}
+	if r.Error != "" {
+		return nil, -1, fmt.Errorf(r.Error)
 	}
+	return r.SqlStatements, int(r.Index), nil
+}
 
-	return pr.Statements, pr.Position, nil
+func getEnv(name, fallback string) string {
```

Comment:
Following my suggestions above, we can remove this function definition.

Reply:
Sure, I will remove this function.